### PR TITLE
Create and add instance reference inside ResolveMetadata*

### LINKED ISSUE #37168: Avoid remembering last metadata to point to $id when AddReference fails
Instead, call `AddReference` when the reader is at the `$id` property to effectively point to it.

### PR DESCRIPTION
**Pull request by jozkee**
...in order to remove workaround described in https://github.com/dotnet/runtime/issues/37168 used to correctly build the JSON path in case of error.

This PR adds a new method `CreateInstance` to `JsonResumableConverter<T>` that is used to instantiate the object at the time that `ReferenceResolver.AddReference` is called.

It also splits `ResolveMetadata` into two methods: `ResolveMetadataForJsonObject` (Used on POCOs and dictionaries) and `ResolveMetadataForJsonArray` (Used for collections such as `List<T>`).

This makes the code more mantainable, reduces the amount of if checks and allow us to remove the string field `ReadStackFrame.MetadataId` from the state.

Fixes https://github.com/dotnet/runtime/issues/37168.

### CODE REVIEW

**Comment by Jozkee** [comment]
Perf. results for the stream scenarios affected by this change:

```
summary:
better: 9, geomean: 1.020
worse: 1, geomean: 1.023
total diff: 10

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Serialization.Tests.ReadJson<LoginViewModel>.DeserializeFromStr |      1.02 |           683.30 |           699.20 |         |

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| -------- |
| System.Text.Json.Serialization.Tests.ReadJson<SimpleStructWithProperties>.Deseri |      1.05 |           532.98 |           505.85 | bimodal |
| System.Text.Json.Serialization.Tests.ReadJson<IndexViewModel>.DeserializeFromStr |      1.02 |         30798.67 |         30088.40 |         |
| System.Text.Json.Serialization.Tests.ReadJson<LargeStructWithProperties>.Deseria |      1.02 |          1533.54 |          1503.43 |         |
| System.Text.Json.Serialization.Tests.ReadJson<ImmutableSortedDictionary<String,  |      1.02 |         75880.01 |         74599.11 |         |
| System.Text.Json.Serialization.Tests.ReadJson<Int32>.DeserializeFromStream       |      1.02 |           262.67 |           258.25 |         |
| System.Text.Json.Serialization.Tests.ReadJson<ImmutableDictionary<String, String |      1.01 |         47830.18 |         47156.48 |         |
| System.Text.Json.Serialization.Tests.ReadJson<HashSet<String>>.DeserializeFromSt |      1.01 |         13382.56 |         13210.33 |         |
| System.Text.Json.Serialization.Tests.ReadJson<Dictionary<String, String>>.Deseri |      1.01 |         22656.52 |         22433.16 |         |
| System.Text.Json.Serialization.Tests.ReadJson<BinaryData>.DeserializeFromStream  |      1.01 |           744.31 |           738.44 |         |
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs [hunk 1]
```diff
@@ -12,17 +12,22 @@ public static partial class JsonSerializer
...
         /// <summary>
         /// Returns true if successful, false is the reader ran out of buffer.
-        /// Sets state.Current.ReturnValue to the $ref target for MetadataRefProperty cases.
+        /// Sets state.Current.ReturnValue to the reference target for $ref cases;
+        /// Sets state.Current.ReturnValue to a new instance for $id cases.
         /// </summary>
-        internal static bool ResolveMetadata(
-            JsonConverter converter,
+        internal static bool ResolveMetadataForJsonObject<T>(
+            JsonResumableConverter<T> converter,
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs:27]: Does this need to be generic? Can we put the virtual method on `JsonConverter`? Same below.

Comment (Jozkee) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs]: Done, thanks for the suggestion.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs [hunk 2]
```diff
@@ -12,18 +12,24 @@ public static partial class JsonSerializer
...
         /// </summary>
-        internal static bool ResolveMetadata(
-            JsonConverter converter,
+        internal static bool ResolveMetadataForJsonObject(
             ref Utf8JsonReader reader,
-            ref ReadStack state)
+            ref ReadStack state,
+            JsonSerializerOptions options)
         {
+            JsonConverter converter = state.Current.JsonClassInfo.PropertyInfoForClassInfo.ConverterBase;
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs:31]: Removed `JsonConverter` from the params to use this one instead.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs [hunk 3]
```diff
@@ -297,7 +297,7 @@ TKey ReadDictionaryKey(ref Utf8JsonReader reader, ref ReadStack state)
             return success;
         }
 
-        public sealed override void CreateInstance(ref Utf8JsonReader reader, ref ReadStack state, JsonSerializerOptions options)
+        internal sealed override void CreateInstanceForReferenceResolver(ref Utf8JsonReader reader, ref ReadStack state, JsonSerializerOptions options)
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs:300]: The type is already internal, so this should be left as public - here and elsewhere.

**Review by layomia — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs (hunk 3 above)
Comment (Jozkee) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs:300]: This method is inherited from `JsonConverter` which is public, if I make it public here and in the base, then we would be adding new API to the base type.

**Pull request merged**